src/groth16/mod.rs

```diff
@@ -466,6 +771,170 @@ impl<'a, E: Engine> ParameterSource<E> for &'a Parameters<E> {
     }
 }
 
+// A re-usable method for parameter loading via mmap.
+fn read_g1<E: Engine>(
```

Comment:
any reason these are not used above?

Reply:
A comment was added, but in short those are used to pull in the params when access at already known locations is needed and don't update any offset state.  Some consolidation of those methods and the internal closures could be done if it's a priority.